drivers/video/arducam_mega.c

```diff
@@ -0,0 +1,1048 @@
+{
+	const struct arducam_mega_config *cfg = dev->config;
+	struct arducam_mega_data *drv_data = dev->data;
+	uint8_t tries = 200;
+
+	arducam_mega_write_reg(&cfg->bus, ARDUCHIP_FIFO, FIFO_CLEAR_ID_MASK);
+	arducam_mega_write_reg(&cfg->bus, ARDUCHIP_FIFO, FIFO_START_MASK);
+
+	do {
+		if (tries-- == 0) {
+			LOG_ERR("Capture timeout!");
+			return -1;
+		}
+		k_msleep(2);
+	} while (!(arducam_mega_read_reg(&cfg->bus, ARDUCHIP_TRIG) & CAP_DONE_MASK));
+
+	drv_data->fifo_length = arducam_mega_read_reg(&cfg->bus, FIFO_SIZE1);
+	drv_data->fifo_length |= (arducam_mega_read_reg(&cfg->bus, FIFO_SIZE2) << 8);
+	drv_data->fifo_length |= (arducam_mega_read_reg(&cfg->bus, FIFO_SIZE3) << 16);
+
+	drv_data->fifo_first_read = 1;
+	*length = drv_data->fifo_length;
+	return 0;
+}
+
+static int arducam_mega_fifo_read(const struct device *dev, struct video_buffer *buf)
+{
+	int ret;
+	int32_t rlen;
+	const struct arducam_mega_config *cfg = dev->config;
+	struct arducam_mega_data *drv_data = dev->data;
+
+	rlen = buf->size > drv_data->fifo_length ? drv_data->fifo_length : buf->size;
+
+	LOG_DBG("read fifo :%u. - fifo_length %u", buf->size, drv_data->fifo_length);
+
+	ret = arducam_mega_read_block(&cfg->bus, buf->buffer, rlen, drv_data->fifo_first_read);
+
+	if (ret == 0) {
+		drv_data->fifo_length -= rlen;
+		buf->bytesused = rlen;
+		if (drv_data->fifo_first_read) {
+			drv_data->fifo_first_read = 0;
+		}
+	}
+
+	return ret;
+}
+
+static void __buffer_work(struct k_work *work)
+{
+	struct k_work *dwork = work;
+	struct arducam_mega_data *drv_data =
+		CONTAINER_OF(dwork, struct arducam_mega_data, buf_work);
+	static uint32_t f_timestamp, f_length;
+	struct video_buffer *vbuf;
+
+	vbuf = k_fifo_get(&drv_data->fifo_in, K_FOREVER);
+
+
+	if (vbuf == NULL) {
+		return;
+	}
+
+	if (drv_data->fifo_length == 0) {
+		arducam_mega_capture(drv_data->dev, &f_length);
+		f_timestamp = k_uptime_get_32();
+	}
+
+	arducam_mega_fifo_read(drv_data->dev, vbuf);
+
+	if (drv_data->fifo_length == 0) {
+		vbuf->flags = VIDEO_BUF_EOF;
+	} else {
+		vbuf->flags = VIDEO_BUF_FRAG;
+		k_work_submit_to_queue(&ac_work_q, &drv_data->buf_work);
+	}
+
+	vbuf->timestamp = f_timestamp;
+	vbuf->bytesframe = f_length;
+	k_fifo_put(&drv_data->fifo_out, vbuf);
+
+	k_yield();
+}
+
+static int arducam_mega_enqueue(const struct device *dev, enum video_endpoint_id ep,
+				struct video_buffer *vbuf)
+{
+	struct arducam_mega_data *data = dev->data;
+
+	if (ep != VIDEO_EP_OUT) {
+		return -EINVAL;
+	}
+	k_fifo_put(&data->fifo_in, vbuf);
+
+	LOG_DBG("enqueue buffer %p", vbuf->buffer);
+
+	return 0;
+}
+
+static int arducam_mega_dequeue(const struct device *dev, enum video_endpoint_id ep,
+				struct video_buffer **vbuf, k_timeout_t timeout)
+{
+	struct arducam_mega_data *data = dev->data;
+
+	if (ep != VIDEO_EP_OUT) {
+		return -EINVAL;
+	}
+
+	*vbuf = k_fifo_get(&data->fifo_out, timeout);
+
+	LOG_DBG("dequeue buffer %p", (*vbuf)->buffer);
+
```

Comment:
It is not possible to keep this line before `vbuf == NULL` as `(*vbuf)->buffer` would trigger a fault.

It can be put after, or even removed altogether as you prefer.